nupic/research/frameworks/pytorch/restore_utils.py

```diff
@@ -0,0 +1,147 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2018, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+import io
+import os
+import pickle
+
+import torch
+
+
+def get_state_dict(checkpoint_path):
+
+    checkpoint_path = os.path.expanduser(checkpoint_path)
+    with open(checkpoint_path, "rb") as loaded_state:
+        checkpoint_dict = pickle.load(loaded_state)
+
+    if "model" in checkpoint_dict:
+        with io.BytesIO(checkpoint_dict["model"]) as buffer:
+            state_dict = torch.load(buffer)
+        return state_dict
+    else:
+        return None
+
+
+def get_linear_param_names(model):
+
+    linear_params = []
+    for name_m, m in model.named_modules():
+        if isinstance(m, torch.nn.Linear):
+            for name_p, _ in m.named_parameters():
+                full_name = name_m + ("." if name_m else "") + name_p
+                linear_params.append(full_name)
+    return linear_params
+
+
+def load_multi_state(model, config):
+    """
+    Example 1:
+    ```
+    checkpoint_linear = "~/.../checkpoint_20/checkpoint"
+    checkpoint_nonlinear = "~/.../checkpoint_1/checkpoint""
+
+    config = {
+        "restore_linear": checkpoint_linear,
+        "restore_nonlinear": checkpoint_nonlinear,
+    }
+
+    model = ResNet()
+    model = load_multi_state(model, config)
+    ```
+
+    Example 2:
+    ```
+    checkpoint_model = "~/.../checkpoint_1/checkpoint""
+
+    config = {
+        "restore_full_model": checkpoint_model,
+    }
+
+    model = ResNet()
+    model = load_multi_state(model, config)
+    ```
+    """
+
+    # Identify checkpoints for the model and it's sub-components.
+    model_source = config.get("restore_full_model", None)
+    linear_source = config.get("restore_linear", None)
+    nonlinear_source = config.get("restore_nonlinear", None)
```

Comment:
Dont use a dict as single argument to a function like this. 
It is much easier to read and maintain when the function arguments are passed as individual arguments with default values and documented in the docstring.

Reply:
Fair point